docs/src/developers/style.md

```diff
@@ -51,6 +51,220 @@ format("test")
 The following sections outline extra style guide points that are not fixed
 automatically by JuliaFormatter.
 
+### Abstract types and composition
+
+Specifying types for method arguments and struct fields is mostly optional in
+Julia. The benefit of abstract method arguments is that it enables functions
+and types from one package to be used with functions and types from another
+package via multiple dispatch.
+
+However, abstractly typed methods have two main drawbacks:
+
+ 1. It's possible to find out that you are working with unexpected types deep
+    in the call chain, potentially leading to hard-to-diagnose [`MethodError`s](https://docs.julialang.org/en/v1/manual/methods/#Defining-Methods).
+ 2. Untyped function arguments can lead to correctness problems if the user's
+    choice of input type does not satisfy the assumptions made by the author of
+    the function.
+
+As a motivating example, consider the following function:
+```jldoctest my_sum
+julia> function my_sum(x)
+           y = 0.0
+           for i in 1:length(x)
+               y += x[i]
+           end
+           return y
+       end
+my_sum (generic function with 1 method)
+```
+This function contains a number of implicit assumptions about the type of `x`:
+ * `x` supports 1-based `getindex` and implements `length`
+ * The element type of `x` supports addition with `0.0`, and then with the
+   result of `x + 0.0`.
+
+!!! info
+    As a motivating example for the second point, [`VariableRef`](@ref) plus
+    `Float64` produces an [`AffExpr`](@ref). Do not assume that `+(::T, ::T)`
+    produces the type `T`.
```

Comment:
```suggestion
    `Float64` produces an [`AffExpr`](@ref). Do not assume that `+(::A, ::B)`
    produces the type `A` or `B`.
```
If you want to prove the point for a same type `T`, another example would be more appropriate